File: app/src/main/java/com/stevenschoen/putionew/PutioActivity.java
package com.stevenschoen.putionew;

import android.app.SearchManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.animation.FastOutSlowInInterpolator;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;

import com.aurelhubert.ahbottomnavigation.AHBottomNavigation;
import com.aurelhubert.ahbottomnavigation.AHBottomNavigationItem;
import com.stevenschoen.putionew.cast.BaseCastActivity;
import com.stevenschoen.putionew.files.FilesFragment;
import com.stevenschoen.putionew.fragments.Account;
import com.stevenschoen.putionew.model.files.PutioFile;
import com.stevenschoen.putionew.model.transfers.PutioTransfer;
import com.stevenschoen.putionew.transfers.AddTransferActivity;
import com.stevenschoen.putionew.transfers.Transfers;
import com.stevenschoen.putionew.tv.TvActivity;

import org.apache.commons.io.FileUtils;
import org.jetbrains.annotations.NotNull;

import java.io.File;

public class PutioActivity extends BaseCastActivity {

    public static final int TAB_ACCOUNT = 0;
    public static final int TAB_FILES = 1;
    public static final int TAB_TRANSFERS = 2;

	public static final String FRAGTAG_ACCOUNT = "account";
	public static final String FRAGTAG_FILES = "files";
	public static final String FRAGTAG_TRANSFERS = "transfers";

    private boolean init = false;

    private SharedPreferences sharedPrefs;

    public static final String checkCacheSizeIntent = "com.stevenschoen.putionew.checkcachesize";
    public static final String noNetworkIntent = "com.stevenschoen.putionew.nonetwork";

	private AHBottomNavigation bottomNavView;

    private View addTransferView;
	private boolean showingAddTransferFab = true;

	@Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (UIUtils.isTV(this)) {
            Intent tvIntent = new Intent(this, TvActivity.class);
            startActivity(tvIntent);
            finish();
			return;
        }

        PutioApplication application = (PutioApplication) getApplication();
        if (application.isLoggedIn()) {
            init(savedInstanceState);
        } else {
            Intent setupIntent = new Intent(this, LoginActivity.class);
            startActivity(setupIntent);

            finish();
            return;
        }

        sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);

        if (getIntent() != null) {
            handleIntent(getIntent());
        }

        IntentFilter checkCacheSizeIntentFilter = new IntentFilter(
                PutioActivity.checkCacheSizeIntent);
        IntentFilter noNetworkIntentFilter = new IntentFilter(
                PutioActivity.noNetworkIntent);

        registerReceiver(checkCacheSizeReceiver, checkCacheSizeIntentFilter);
        registerReceiver(noNetworkReceiver, noNetworkIntentFilter);
    }

    @Override
    protected void onNewIntent(Intent intent) {
		super.onNewIntent(intent);

        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        if (init) {
            int goToTab = intent.getIntExtra("goToTab", -1);
            if (goToTab != -1) {
                selectTab(goToTab, true);
            }

            if (intent.getAction() != null) {
                if (intent.getAction().equals(Intent.ACTION_SEARCH) && getFilesFragment() != null) {
                    String query = intent.getStringExtra(SearchManager.QUERY);
                    getFilesFragment().addSearch(query);
                }
            }
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

		outState.putInt("currentTab", bottomNavView.getCurrentItem());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.menu_putio, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_logout:
                logOut();
                return true;
            case R.id.menu_about:
                Intent aboutIntent = new Intent(PutioActivity.this, AboutActivity.class);
                startActivity(aboutIntent);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void init(Bundle savedInstanceState) {
        init = true;

        setContentView(R.layout.main);

		if (savedInstanceState == null) {
			Account accountFragment = (Account) Fragment.instantiate(this, Account.class.getName());
			FilesFragment filesFragment = FilesFragment.Companion.newInstance(this, null);
			Transfers transfersFragment = (Transfers) Fragment.instantiate(this, Transfers.class.getName());
			getSupportFragmentManager().beginTransaction()
					.add(R.id.main_content_holder, accountFragment, FRAGTAG_ACCOUNT)
					.detach(accountFragment)
					.add(R.id.main_content_holder, filesFragment, FRAGTAG_FILES)
					.detach(filesFragment)
					.add(R.id.main_content_holder, transfersFragment, FRAGTAG_TRANSFERS)
					.detach(transfersFragment)
					.commitNow();
		}

        try {
            ((PutioApplication) getApplication()).buildUtils();
        } catch (PutioUtils.NoTokenException e) {
            e.printStackTrace();
        }

		setupLayout();

        int navItem = TAB_FILES;
        if (savedInstanceState != null) {
            navItem = savedInstanceState.getInt("currentTab", -1);
        }
        selectTab(navItem, false);

		addTransferView = findViewById(R.id.main_addtransfer);
		addTransferView.setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(View v) {
				PutioFile destinationFolder = null;
				if (bottomNavView.getCurrentItem() == TAB_FILES) {
					destinationFolder = getFilesFragment().getCurrentPage().getFile();
				}
				Intent addTransferIntent = new Intent(PutioActivity.this, AddTransferActivity.class);
				if (destinationFolder != null) {
					addTransferIntent.putExtra(AddTransferActivity.EXTRA_DESTINATION_FOLDER, destinationFolder);
				}
				startActivity(addTransferIntent);
			}
		});
		updateAddTransferFab(false);
    }

	@Override
	public void onAttachFragment(Fragment fragment) {
		super.onAttachFragment(fragment);
		if (fragment.getTag() != null) {
			switch (fragment.getTag()) {
				case FRAGTAG_FILES:
					((FilesFragment) fragment).setCallbacks(new FilesFragment.Callbacks() {
						@Override
						public void onSelectionStarted() {
							if (init) {
								updateAddTransferFab(true);
							}
						}

						@Override
						public void onSelectionEnded() {
							updateAddTransferFab(true);
						}

						@Override
						public void onCurrentFileChanged() {
							updateAddTransferFab(true);
						}
					});
					break;
				case FRAGTAG_TRANSFERS: {
					((Transfers) fragment).setCallbacks(new Transfers.Callbacks() {
						@Override
						public void onTransferSelected(@NotNull PutioTransfer transfer) {
							showFilesAndHighlightFile(transfer.saveParentId, transfer.fileId);
						}
					});
				}
			}
		}
	}

	private void updateAddTransferFab(boolean animate) {
		boolean shouldShow = shouldShowAddTransferFab();
		if (shouldShow && !showingAddTransferFab) {
			showingAddTransferFab = true;
			if (animate) {
				addTransferView.animate()
						.setInterpolator(new FastOutSlowInInterpolator())
						.alpha(1f)
						.scaleX(1f)
						.scaleY(1f);
			} else {
				addTransferView.setAlpha(1f);
				addTransferView.setScaleX(1f);
				addTransferView.setScaleY(1f);
			}
			addTransferView.setEnabled(true);
			addTransferView.setFocusable(true);
			addTransferView.setClickable(true);
		} else if (!shouldShow && showingAddTransferFab) {
			showingAddTransferFab = false;
			if (animate) {
				addTransferView.animate()
						.setInterpolator(new FastOutSlowInInterpolator())
						.alpha(0f)
						.scaleX(0f)
						.scaleY(0f);
			} else {
				addTransferView.setAlpha(0f);
				addTransferView.setScaleX(0f);
				addTransferView.setScaleY(0f);
			}
			addTransferView.setEnabled(false);
			addTransferView.setFocusable(false);
			addTransferView.setClickable(false);
		}
	}

	private boolean shouldShowAddTransferFab() {
		switch (bottomNavView.getCurrentItem()) {
			case TAB_ACCOUNT: return false;
			case TAB_FILES: {
				FilesFragment filesFragment = getFilesFragment();
				if (filesFragment.isSelecting()) {
					return false;
				} else {
					FilesFragment.Page currentPage = filesFragment.getCurrentPage();
					if (currentPage == null) {
						return true;
					} else {
						if (currentPage.getType() == FilesFragment.Page.Type.Search) {
							return false;
						} else {
							if (currentPage.getFile().isFolder()) {
								return true;
							} else {
								return false;
							}
						}
					}
				}
			}
			case TAB_TRANSFERS: return true;
			default: return true;
		}
	}

	public void logOut() {
        sharedPrefs.edit().remove("token").commit();
        finish();
        startActivity(getIntent());
    }

    private void setupLayout() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.main_toolbar);
        setSupportActionBar(toolbar);

		bottomNavView = (AHBottomNavigation) findViewById(R.id.main_bottom_nav);
		bottomNavView.setDefaultBackgroundColor(Color.parseColor("#F8F8F8"));
		bottomNavView.setAccentColor(Color.BLACK);
		bottomNavView.setInactiveColor(Color.parseColor("#80000000"));
		bottomNavView.addItem(new AHBottomNavigationItem(getString(R.string.account), R.drawable.ic_nav_account));
		bottomNavView.addItem(new AHBottomNavigationItem(getString(R.string.files), R.drawable.ic_nav_files));
		bottomNavView.addItem(new AHBottomNavigationItem(getString(R.string.transfers), R.drawable.ic_nav_transfers));
		bottomNavView.setOnTabSelectedListener(new AHBottomNavigation.OnTabSelectedListener() {
			@Override
			public boolean onTabSelected(int position, boolean wasSelected) {
				bottomNavView.post(new Runnable() {
					@Override
					public void run() {
						updateAddTransferFab(true);
					}
				});
				switch (position) {
					case TAB_ACCOUNT:
					case TAB_FILES:
					case TAB_TRANSFERS: showFragment(position, true); return true;
				}
				return false;
			}
		});
    }

	public void showFilesAndHighlightFile(long parentId, long id) {
        selectTab(TAB_FILES, true);
//        getFilesFragment().highlightFile(parentId, id);
    }

	private Account getAccountFragment() {
		return (Account) getSupportFragmentManager().findFragmentByTag(FRAGTAG_ACCOUNT);
	}

    private FilesFragment getFilesFragment() {
        return (FilesFragment) getSupportFragmentManager().findFragmentByTag(FRAGTAG_FILES);
    }

    private Transfers getTransfersFragment() {
		return (Transfers) getSupportFragmentManager().findFragmentByTag(FRAGTAG_TRANSFERS);
    }

    private BroadcastReceiver checkCacheSizeReceiver = new BroadcastReceiver() {

        @Override
        public void onReceive(Context context, Intent intent) {
            checkCacheSize();
        }
    };

    private BroadcastReceiver noNetworkReceiver = new BroadcastReceiver() {

        @Override
        public void onReceive(Context context, Intent intent) {
            getTransfersFragment().setHasNetwork(false);
        }
    };

	@Override
    public void onBackPressed() {
        if (bottomNavView.getCurrentItem() == TAB_FILES) {
			if (!getFilesFragment().goBack(true)) {
				super.onBackPressed();
			}
		} else {
            super.onBackPressed();
        }
    }

	private void showFragment(int position, boolean animate) {
		FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
		if (animate) {
			ft.setCustomAnimations(R.anim.bottomnav_enter, R.anim.bottomnav_exit);
		}
		switch (position) {
			case TAB_ACCOUNT: {
				ft.attach(getAccountFragment());
				ft.detach(getFilesFragment());
				ft.detach(getTransfersFragment());
			} break;
			case TAB_FILES: {
				ft.detach(getAccountFragment());
				ft.attach(getFilesFragment());
				ft.detach(getTransfersFragment());
			} break;
			case TAB_TRANSFERS: {
				ft.detach(getAccountFragment());
				ft.detach(getFilesFragment());
				ft.attach(getTransfersFragment());
			} break;
		}
		ft.commitNow();
	}

	private void selectTab(int position, boolean animate) {
        if (bottomNavView.getCurrentItem() != position) {
            bottomNavView.setCurrentItem(position, false);
			showFragment(position, animate);
        }
    }

    @Override
    protected void onDestroy() {
		if (init) {
			unregisterReceiver(checkCacheSizeReceiver);
			unregisterReceiver(noNetworkReceiver);
		}

        super.onDestroy();
    }

	@Override
	public Integer getCastMiniControllerContainerId() {
		return R.id.holder_castbar;
	}

	public void checkCacheSize() {
        int maxSize = sharedPrefs.getInt("maxCacheSizeMb", 20);
        File cache = getCacheDir();
        if (FileUtils.sizeOf(cache) >= (FileUtils.ONE_MB * maxSize)) {
            File[] cacheFiles = cache.listFiles();
			for (File file : cacheFiles) {
				if (!file.getName().equals("0")) {
					FileUtils.deleteQuietly(file);
				}
			}
        }
    }
}

File: app/src/main/java/com/stevenschoen/putionew/PutioActivity.kt
package com.stevenschoen.putionew

import android.app.SearchManager
import android.app.job.JobInfo
import android.app.job.JobScheduler
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.stevenschoen.putionew.account.AccountFragment
import com.stevenschoen.putionew.cast.BaseCastActivity
import com.stevenschoen.putionew.files.FileDownloadsMaintenanceService
import com.stevenschoen.putionew.files.FilesFragment
import com.stevenschoen.putionew.files.FolderLoader
import com.stevenschoen.putionew.model.files.PutioFile
import com.stevenschoen.putionew.model.transfers.PutioTransfer
import com.stevenschoen.putionew.transfers.TransfersFragment
import com.stevenschoen.putionew.transfers.add.AddTransferActivity
import com.stevenschoen.putionew.tv.TvActivity
import java.util.concurrent.TimeUnit

class PutioActivity : BaseCastActivity() {

  var init = false

  val sharedPrefs by lazy { PreferenceManager.getDefaultSharedPreferences(this)!! }

  lateinit var bottomNavView: BottomNavigationView

  lateinit var addTransferView: View
  var showingAddTransferFab = true

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    if (UIUtils.isTV(this)) {
      val tvIntent = Intent(this, TvActivity::class.java)
      startActivity(tvIntent)
      finish()
      return
    }

    val application = application as PutioApplication
    if (application.isLoggedIn) {
      init(savedInstanceState)
    } else {
      val setupIntent = Intent(this, LoginActivity::class.java)
      startActivity(setupIntent)

      finish()
      return
    }

    intent?.let(::handleIntent)

    val noNetworkIntentFilter = IntentFilter(
        PutioActivity.noNetworkIntent
    )

    registerReceiver(noNetworkReceiver, noNetworkIntentFilter)

    val jobDispatcher = getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
    if (jobDispatcher.allPendingJobs.none {
          it.id == FileDownloadsMaintenanceService.FILE_DOWNLOADS_MAINTENANCE_JOB_ID
        }) {
      jobDispatcher.schedule(
          JobInfo.Builder(
              FileDownloadsMaintenanceService.FILE_DOWNLOADS_MAINTENANCE_JOB_ID,
              ComponentName(this, FileDownloadsMaintenanceService::class.java)
          )
              .setPeriodic(TimeUnit.DAYS.toMillis(2))
              .setPersisted(true)
              .setRequiresDeviceIdle(true)
              .build()
      )
    }

    jobDispatcher.schedule(
        JobInfo.Builder(
            FileDownloadsMaintenanceService.FILE_DOWNLOADS_MAINTENANCE_JOB_ID,
            ComponentName(this, FileDownloadsMaintenanceService::class.java)
        )
            .setOverrideDeadline(5000)
            .setPersisted(true)
            .setRequiresDeviceIdle(true)
            .build()
    )
  }

  override fun onNewIntent(intent: Intent) {
    super.onNewIntent(intent)

    handleIntent(intent)
  }

  private fun handleIntent(intent: Intent) {
    if (init) {
      val goToTab = intent.getIntExtra(EXTRA_GO_TO_TAB, -1)
      intent.removeExtra(EXTRA_GO_TO_TAB)
      if (goToTab != -1) {
        selectTab(goToTab, true)
      }

      if (intent.action != null) {
        if (intent.action == Intent.ACTION_SEARCH) {
          val query = intent.getStringExtra(SearchManager.QUERY)
          filesFragment.addSearch(query)
        }
      }
    }
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    super.onCreateOptionsMenu(menu)
    menuInflater.inflate(R.menu.menu_putio, menu)

    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    when (item.itemId) {
      R.id.menu_logout -> {
        logOut()
        return true
      }
      R.id.menu_about -> {
        val aboutIntent = Intent(this@PutioActivity, AboutActivity::class.java)
        startActivity(aboutIntent)
        return true
      }
    }

    return super.onOptionsItemSelected(item)
  }

  private fun init(savedInstanceState: Bundle?) {
    init = true

    setContentView(R.layout.main)

    setupLayout()

    if (savedInstanceState == null) {
      val accountFragment = Fragment.instantiate(this, AccountFragment::class.java.name) as AccountFragment
      val filesFragment = FilesFragment.newInstance(this, null)
      val transfersFragment = Fragment.instantiate(this, TransfersFragment::class.java.name) as TransfersFragment
      supportFragmentManager.beginTransaction()
          .add(R.id.main_content_holder, accountFragment, FRAGTAG_ACCOUNT)
          .detach(accountFragment)
          .add(R.id.main_content_holder, filesFragment, FRAGTAG_FILES)
          .detach(filesFragment)
          .add(R.id.main_content_holder, transfersFragment, FRAGTAG_TRANSFERS)
          .detach(transfersFragment)
          .commitNow()

      selectTab(TAB_FILES, false)
    }

    addTransferView = findViewById(R.id.main_addtransfer)
    addTransferView.setOnClickListener {
      var destinationFolder: PutioFile? = null
      if (bottomNavView.selectedItemId == TAB_FILES) {
        destinationFolder = (filesFragment.currentPage as FilesFragment.Page.File).file
      }
      val addTransferIntent = Intent(this@PutioActivity, AddTransferActivity::class.java)
      if (destinationFolder != null) {
        addTransferIntent.putExtra(AddTransferActivity.EXTRA_PRECHOSEN_DESTINATION_FOLDER, destinationFolder)
      }
      startActivity(addTransferIntent)
    }
    updateAddTransferFab(false)
  }

  override fun onAttachFragment(fragment: Fragment?) {
    super.onAttachFragment(fragment)
    if (fragment!!.tag != null) {
      when (fragment.tag) {
        FRAGTAG_FILES -> (fragment as FilesFragment).callbacks = object : FilesFragment.Callbacks {
          override fun onSelectionStarted() {
            if (init) {
              updateAddTransferFab(true)
            }
          }

          override fun onSelectionEnded() {
            updateAddTransferFab(true)
          }

          override fun onCurrentFileChanged() {
            updateAddTransferFab(true)
          }
        }
        FRAGTAG_TRANSFERS -> {
          (fragment as TransfersFragment).callbacks = object : TransfersFragment.Callbacks {
            override fun onTransferSelected(transfer: PutioTransfer) {
              showFilesAndGoToFile(transfer.saveParentId, transfer.fileId)
            }
          }
        }
      }
    }
  }

  private fun updateAddTransferFab(animate: Boolean) {
    val shouldShow = shouldShowAddTransferFab()
    if (shouldShow && !showingAddTransferFab) {
      showingAddTransferFab = true
      if (animate) {
        addTransferView.animate()
            .setInterpolator(FastOutSlowInInterpolator())
            .alpha(1f)
            .scaleX(1f)
            .scaleY(1f)
      } else {
        addTransferView.alpha = 1f
        addTransferView.scaleX = 1f
        addTransferView.scaleY = 1f
      }
      addTransferView.isEnabled = true
      addTransferView.isFocusable = true
      addTransferView.isClickable = true
    } else if (!shouldShow && showingAddTransferFab) {
      showingAddTransferFab = false
      if (animate) {
        addTransferView.animate()
            .setInterpolator(FastOutSlowInInterpolator())
            .alpha(0f)
            .scaleX(0f)
            .scaleY(0f)
      } else {
        addTransferView.alpha = 0f
        addTransferView.scaleX = 0f
        addTransferView.scaleY = 0f
      }
      addTransferView.isEnabled = false
      addTransferView.isFocusable = false
      addTransferView.isClickable = false
    }
  }

  private fun shouldShowAddTransferFab(): Boolean {
    return when (bottomNavView.selectedItemId) {
      TAB_ACCOUNT -> false
      TAB_FILES -> {
        val filesFragment = filesFragment
        if (filesFragment.isSelecting) {
          false
        } else {
          val currentPage = filesFragment.currentPage
          when (currentPage) {
            null -> true
            is FilesFragment.Page.Search -> false
            is FilesFragment.Page.File -> currentPage.file.isFolder
          }
        }
      }
      TAB_TRANSFERS -> true
      else -> true
    }
  }

  fun logOut() {
    (application as PutioApplication).putioUtils = null
    sharedPrefs!!.edit().remove("token").apply()
    FolderLoader.DiskCache(this).deleteCache()
    finish()
    startActivity(intent)
  }

  private fun setupLayout() {
    val toolbar = findViewById<Toolbar>(R.id.main_toolbar)
    setSupportActionBar(toolbar)

    bottomNavView = findViewById(R.id.main_bottom_nav)
    bottomNavView.apply {
      menu.add(0, TAB_ACCOUNT, 0, R.string.account).apply {
        setIcon(R.drawable.ic_nav_account)
      }
      menu.add(0, TAB_FILES, 0, R.string.files).apply {
        setIcon(R.drawable.ic_nav_files)
      }
      menu.add(0, TAB_TRANSFERS, 0, R.string.transfers).apply {
        setIcon(R.drawable.ic_nav_transfers)
      }
      setOnNavigationItemSelectedListener { item ->
        post { updateAddTransferFab(true) }
        showFragment(item.itemId, true)
        true
      }
      setOnNavigationItemReselectedListener { item ->
        if (item.itemId == TAB_FILES) {
          filesFragment.goBackToRoot()
        }
      }
    }
  }

  fun showFilesAndGoToFile(parentId: Long, id: Long) {
    selectTab(TAB_FILES, true)
    filesFragment.goToFile(parentId, id)
  }

  private val accountFragment: AccountFragment
    get() = supportFragmentManager.findFragmentByTag(FRAGTAG_ACCOUNT) as AccountFragment

  private val filesFragment: FilesFragment
    get() = supportFragmentManager.findFragmentByTag(FRAGTAG_FILES) as FilesFragment

  private val transfersFragment: TransfersFragment
    get() = supportFragmentManager.findFragmentByTag(FRAGTAG_TRANSFERS) as TransfersFragment

  private val noNetworkReceiver = object : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
      transfersFragment.setHasNetwork(false)
    }
  }

  override fun onBackPressed() {
    if (bottomNavView.selectedItemId == TAB_FILES) {
      if (!filesFragment.goBack(true)) {
        super.onBackPressed()
      }
    } else {
      super.onBackPressed()
    }
  }

  private fun showFragment(position: Int, animate: Boolean) {
    val transaction = supportFragmentManager.beginTransaction()
    if (animate) {
      transaction.setCustomAnimations(R.animator.bottomnav_enter, R.animator.bottomnav_exit)
    }
    when (position) {
      TAB_ACCOUNT -> {
        transaction.attach(accountFragment)
        transaction.detach(filesFragment)
        transaction.detach(transfersFragment)
      }
      TAB_FILES -> {
        transaction.detach(accountFragment)
        transaction.attach(filesFragment)
        transaction.detach(transfersFragment)
      }
      TAB_TRANSFERS -> {
        transaction.detach(accountFragment)
        transaction.detach(filesFragment)
        transaction.attach(transfersFragment)
      }
    }
    transaction.commitNow()
  }

  private fun selectTab(position: Int, animate: Boolean) {
    if (bottomNavView.selectedItemId != position) {
      bottomNavView.selectedItemId = position
      showFragment(position, animate)
    }
  }

  override fun onDestroy() {
    if (init) {
      unregisterReceiver(noNetworkReceiver)
    }

    super.onDestroy()
  }

  override val castMiniControllerContainerId = R.id.holder_castbar

  companion object {
    const val EXTRA_GO_TO_TAB = "go_to_tab"

    const val TAB_ACCOUNT = 0
    const val TAB_FILES = 1
    const val TAB_TRANSFERS = 2

    private const val FRAGTAG_ACCOUNT = "account"
    private const val FRAGTAG_FILES = "files"
    private const val FRAGTAG_TRANSFERS = "transfers"

    const val noNetworkIntent = "com.stevenschoen.putionew.nonetwork"
  }

}
